File: tests/dynamic_fixtures/rust/sqli_positive.rs
/// SQL injection — positive fixture.
///
/// Vulnerable function: directly concatenates user input into SQL.
/// Expected verdict: Confirmed (UNION payload causes "NYX_SQL_CONFIRMED" in output).
/// Cap: SQL_QUERY  Entry: `run(payload: &str)`
pub fn run(payload: &str) {
    use rusqlite::Connection;

    let conn = Connection::open_in_memory().expect("open in-memory db");
    conn.execute_batch(
        "CREATE TABLE users (id INTEGER, name TEXT);\
         INSERT INTO users VALUES (1, 'alice');\
         INSERT INTO users VALUES (2, 'bob');",
    )
    .expect("setup schema");

    // Vulnerable: direct string concatenation into SQL.
    let query = format!("SELECT name FROM users WHERE name='{}'", payload);

    // Sentinel: the sink (conn.prepare) is reachable with tainted input.
    println!("__NYX_SINK_HIT__");
    let _ = std::io::Write::flush(&mut std::io::stdout());

    match conn.prepare(&query) {
        Ok(mut stmt) => {
            let _ = stmt.query_map([], |row| row.get::<_, String>(0)).map(|rows| {
                for name in rows.flatten() {
                    println!("{}", name);
                }
            });
        }
        Err(e) => {
            // Error-based: print query on failure (oracle can detect via query echo).
            println!("DB query: {}", query);
            println!("DB error: {}", e);
        }
    }
}

File: tests/dynamic_fixtures/python/sqli_positive.py
"""SQL injection — positive fixture.

Vulnerable function: directly concatenates user input into SQL.
Expected verdict: Confirmed (SQLI corpus UNION payload causes "NYX_SQL_CONFIRMED"
to appear in output when the fixture prints query results).
"""
import sqlite3


def login(username):
    """Vulnerable login: direct string concatenation in SQL query."""
    conn = sqlite3.connect(":memory:")
    conn.execute("CREATE TABLE users (id INTEGER, name TEXT)")
    conn.execute("INSERT INTO users VALUES (1, 'alice')")
    conn.execute("INSERT INTO users VALUES (2, 'bob')")
    # Vulnerable: user input directly concatenated
    query = "SELECT name FROM users WHERE name='" + username + "'"
    try:
        rows = conn.execute(query).fetchall()
        for row in rows:
            print(row[0])
    except sqlite3.OperationalError as e:
        # Error-based: print query on failure (common in debug mode)
        print(f"DB query: {query}")
        print(f"DB error: {e}", end="")
    finally:
        conn.close()
